pulsar-common/src/main/proto/PulsarApi.proto

```diff
@@ -61,6 +61,9 @@ message MessageIdData {
     optional int32 batch_index = 4 [default = -1];
     repeated int64 ack_set = 5;
     optional int32 batch_size = 6;
+
+    // For the chunk message id, we need to specify the first chunk message id.
+    optional MessageIdData first_chunk_message_id = 7;
```

Comment:
Since ChunkMessageID serialization and deserialization depend on the `MessageIdData` defined by the proto, we need to change it here even if it does not need to be passed to the server-side.

Reply:
this seems a fundamental issue and it doesn't solve the actual problem. you need messageId `(entryId, ledgerId)` of the first chunk to be present into every chunk of entire message so, you can know what's the first messageId among all chunks when you seek the message. and in this PR you are trying to set that first chunk's messageId (entryId, ledgerId) into every chunk on `ProducerImpl::ackReceived` which doesn't work because you are considering best case where you will receive the first chunk's ack before you send other chunks and that's not guaranteed. Becasuse if you have 10 chunks which will be individual messages and you publish them async and you will get ack of first chunk's message after you publish all other 9 chunks, in that case none of the chunk will have `first_chunk_message_id` and it will not work as Pulsar will claim it should work. 
This is a tricky to solve at client side because we can not update message metadata based on previous message's ack from broker unless you block the publish of other 9 chunks until you receive ack from first chunk, but that will highly impact publish performance and not usable feature.

Reply:
When we send chunk messages, each chunk is sent in order, so I don't think there is a case where the last chunk  be acked before the first chunk.

In the original logic, we determine whether the whole message is sent successfully by receiving the ack of the last chunk, and then return last chunk message id to the user instead of returning message ids of all chunks to the user. 
When we receive the ack of the last chunk, it actually proves that the whole message has been sent successfully and the message id of the first chunk will not be null. In this PR, the first chunk message id and last chunk message id will be packaged into ChunkMessageIdImpl and returned to the user only when the ack of the last chunk is received.

Reply:
> each chunk is sent in order, so I don't think there is a case where the last chunk will be acked before the first chunk.


The issue is not ordering of ack but the timing of ack. if you don't receive ack of first-chunk while publishing other chunks then you can't update `first_chunk_message_id` in rest of the chunks. if chunks won't have `first_chunk_message_id` then when the consumer receives chunked-message, those chunks won't have `firstChunkMessageId` in the metadata and consumer won't be able to seek to first messageId. which will defeat purpose of this PR.

Reply:
> if you don't receive ack of first-chunk while publishing other chunks then you can't update first_chunk_message_id in rest of the chunks.

The ack mentioned here is the ack from the broker and not the ack from the consumer, right? The `ProducerImpl-> ackReceived ` is handling the ack from the broker side.  And we don't need to receive ack of the first-chunk before publishing other chunks. We only need to process the ack(from the broker), and when we processing to the last chunk, the first chunk message id must be present. 

> if chunks won't have first_chunk_message_id then when the consumer receives chunked-message, those chunks won't have firstChunkMessageId in the metadata and consumer won't be able to seek to first messageId.

The producer does not send the first chunk message id to the consumer. The ChunkMessageId returned to the user by the consumer is generated by [getting the ChunkedMessageCtx->chunkedMessageIds](https://github.com/apache/pulsar/pull/12403/files#diff-f6e4c1c4091aa10525f331e48e66b29f22b9f7987755c1b4fb887e24f198bed6R1210-R1211). The Consumer does not access the `first_chunk_message_id `in the `MessageIdData` during this process. 
Note that the `optional MessageIdData first_chunk_message_id = 7;` here is only used in the serialization and deserialization of the ChunkMessageIdImpl but not in the communication between the client and the broker.

Reply:
> and when we processing to the last chunk, the first chunk message id must be present.

it's not guaranteed.

Reply:
> The Consumer does not access the first_chunk_message_id in the MessageIdData during this process.

Please check my below comment in ConsumerImpl class where it accesses it and that's the purpose of this PR to access `first_chunk_message_id` and seek to that messageId. but it's not guaranteed that `first_chunk_message_id` will be present and pulsar can claim this feature incorrectly.

Reply:
@rdhabalia This one is only for the `MessageId.toByteArray()` method, it is a little tricky to add to the wire protocol, but currently we are using the `MessageIdData` to generate the byte array for users, it's will not pass to the broker side let the broker know which one is the first chunk message ID.

Users can store the byte array of the chunk message ID, so that they can recover a chunk message ID by the byte array. If they want to seek to this chunk message continues to consume data with inclusive mode, they don't need to do lots of staff.

And introducing the new chunk message ID will allow us to peek the message from the topic if the application recorded the chunk message ID.

Reply:
@merlimat Please help confirm this part.

Reply:
> And introducing the new chunk message ID will allow us to peek the message from the topic if the application recorded the chunk message ID.

@codelipenghui issue is not with introducing chunk message Id or changes at `MessageId`. the fundamental issue is this entire change is related get `messageId` of first chunk into last chunk data. but that's not guaranteed, so, what's the point of changes which is built on wrong assumption.

Reply:
@rdhabalia No, we don't want to set the first chunk message ID into the last chunk data, just return a chunk message ID to users. Users can persistent the chunk message ID to the state storage, and if they want to start to consume from the stored chunk message, we are able to provide the ability. Otherwise, we will always skip this message, start from the next chunk or normal message.